src/Data/Singletons/Partition.hs

```diff
@@ -220,3 +234,60 @@ isStockOrDefault :: Maybe DerivStrategy -> Bool
 isStockOrDefault Nothing              = True
 isStockOrDefault (Just StockStrategy) = True
 isStockOrDefault (Just _)             = False
+
+buildDataDType :: DsMonad q => Name -> [DTyVarBndr] -> DKind -> q DType
+buildDataDType n tvbs k = do
+  extra_tvbs <- mkExtraDKindBinders k
+  let all_tvbs = tvbs ++ extra_tvbs
+  pure $ foldTypeTvbs (DConT n) all_tvbs
+
+-- | Is this data type a /proper/ GADT? On the surface, this might seem like a
+-- strange question to ask, since @th-desugar@ desugars all data types to GADT
+-- syntax. But there is a subset of GADTs that cannot be expressed as an
+-- equivalent data type using Haskell98 syntax. For instance, this GADT:
+--
+-- @
+-- data Foo :: Type -> Type where
+--   MkFoo :: forall a. a -> Foo a
+-- @
+--
+-- Is equivalent to this Haskell98 data type:
+--
+-- @
+-- data Foo a = MkFoo a
+-- @
+--
+-- However, the following example is a proper GADT:
+--
+-- @
+-- data Bar :: Type -> Type where
+--   MkBar :: Int -> Bar Int
+-- @
+--
+-- Since there is no equivalent Haskell98 data type. The closest you could get
+-- is this:
+--
+-- @
+-- data Bar a = (a ~ Int) => MkBar Int
+-- @
+--
+-- Which requires language extensions to write.
+--
+-- A data type is a proper GADT if one of the following conditions are met:
+--
+-- 1. A constructor has any existentially quantified type variables.
+--
+-- 2. A constructor has a context.
+--
+-- We care about this because some derivable stock classes, such as 'Enum',
+-- forbid derived instances for proper GADTs.
+isProperGADT :: forall q. DsMonad q => DType -> [DCon] -> q Bool
```

Comment:
This is nit-picking, but this function doesn't detect proper GADTs, according to my definition of GADT. For example, it says "yes" to a type with existentials. I think GHC uses the term non-vanilla for these. I suppose it would be more evocative to call them "chocolate" types. Indeed, I might start using this terminology more! Why talk about using "fancy" types or "rich" types or "finely specified" types? Talk about using chocolate types, or even adding some chocolate sauce. Example usage: "What are you so busy doing?" "I'm refactoring my interpreter to add more chocolate sauce. Now invalid constructions will be eliminated by construction."

Reply:
I'm unclear as to how you're distinguishing proper GADTs, non-vanilla data types, and chocolate data types. Is there an ordering on these somehow? Can you give examples of each, and examples of one that isn't an example of the other?

Reply:
`data Ex = forall a. MkEx a` is not vanilla, but is also not a GADT. All proper GADTs are not vanilla, but some non-GADTs are also not vanilla. (I've defined chocolate = not vanilla.)

Reply:
Very well. I've changed `isProperGADT` to be named `isNonVanillaDataType`, to follow existing GHC nomenclature :)